### PR TITLE
[#175126219] Healthcheck step in pipeline

### PR DESCRIPTION
**Pull request by balanza**
This PR introduces an healthcheck step just before promoting the staging slot to production. 

### Behaviour
* deploy application to Staging slot
* perform an http request against `/info` of the staging slot
* if the request is ok (code == 200)
  * switch slot with production
* else
  * abort the pipeline

### Implementation
The implementation relies on [this common template](https://github.com/pagopa/azure-pipeline-templates/tree/master/templates/rest-healthcheck). The important concept is the http request is performed by a separate container which needs some configuration, please refer to the template repository for further info.

### CODE REVIEW

**Review by gunzip — DISMISSED** [review]

**Comment by infantesimone** [comment]
maybe _resources_ needs to be defined in _azure-pipelines.yml_?

**Comment by balanza** [comment]
> maybe _resources_ needs to be defined in _azure-pipelines.yml_?

Good point, I try

File: azure-templates/deploy-steps.yml [hunk 1]
```diff
@@ -26,7 +26,33 @@ parameters:
...
+
+  # min container instance cpu
+  - name: 'containerInstanceCpu'
+    type: string
+    default: '0.5'
+
+  # min container instance memory
+  - name: 'containerInstanceMemory'
+    type: string
+    default: '0.5'
```
Comment (pasqualedevita) [inline azure-templates/deploy-steps.yml:42]: default value for `rest-healthcheck/template.yaml` we can remove it?

File: azure-templates/deploy-steps.yml [hunk 2]
```diff
@@ -98,6 +124,17 @@ steps:
...
+    - template: templates/rest-healthcheck/template.yaml@pagopa-commons 
+      parameters:
+        azureSubscription: '${{ parameters.azureSubscription }}'
+        appName: '${{ parameters.appName }}'
+        endpoint: '${{ parameters.healthcheckEndpoint }}'
+        containerInstanceResourceGroup: '${{ parameters.containerInstanceResourceGroup }}'
+        containerInstanceVNet: '${{ parameters.containerInstanceVNet }}'
+        containerInstanceSubnet: '${{ parameters.containerInstanceSubnet }}'
+        containerInstanceCpu: '${{ parameters.containerInstanceCpu }}'
+        containerInstanceMemory: '${{ parameters.containerInstanceMemory }}'
```
Comment (pasqualedevita) [inline azure-templates/deploy-steps.yml:137]: default value for `rest-healthcheck/template.yaml` we can remove it?

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml [hunk 3]
```diff
@@ -200,6 +207,10 @@ stages:
               azureSubscription: '$(TEST_AZURE_SUBSCRIPTION)'
               resourceGroupName: '$(TEST_RESOURCE_GROUP_NAME)'
               appName: '$(TEST_APP_NAME)'
+              healthcheckEndpoint: 'https://$(PRODUCTION_APP_NAME).staging.azurewebsites.net/info'
```
Comment (pasqualedevita) [inline azure-pipelines.yml:210]:
```suggestion
              healthcheckEndpoint: 'https://$(TEST_APP_NAME).staging.azurewebsites.net/info'
```

File: azure-templates/deploy-steps.yml (hunk 1 above)
Comment (balanza) [reply azure-templates/deploy-steps.yml:42]: ok

File: azure-templates/deploy-steps.yml (hunk 2 above)
Comment (balanza) [reply azure-templates/deploy-steps.yml:137]: ok

**Review by pasqualedevita — COMMENTED** [review]
We can set these values inside the pipeline as a variable? They are not secret and will be more clean what pipeline do

HEALTHCHECK_CONTAINER_RESOURCE_GROUP
HEALTHCHECK_CONTAINER_VNET
HEALTHCHECK_CONTAINER_SUBNET

**Comment by pasqualedevita** [comment]
@gunzip @balanza this function has public access so if `/info` endpoint is exposed
https://github.com/pagopa/io-functions-assets/blob/175126219-healthcheck-pipeline/Info/function.json#L4

**Comment by balanza** [comment]
> @gunzip @balanza this function has public access so if `/info` endpoint is exposed
> https://github.com/pagopa/io-functions-assets/blob/175126219-healthcheck-pipeline/Info/function.json#L4

In healthy conditions, I see no problem. We might consider to obscure error reports in case of errror.

> Container instance and vnet are not necessary

Does the template allow empty vnet?

**Comment by pasqualedevita** [comment]
if we create a switch in the template with simple rest command and with authentication? we can also create a second template

**Comment by balanza** [comment]
For now, id vnet can be there and it's just unnecessary, I'd just leave it as it is.

**Comment by gunzip** [comment]
lgtm this can be deployed only once autoswap on function app service is disabled via terraform I guess cc @uolter

**Comment by uolter** [comment]
> lgtm this can be deployed only once autoswap on function app service is disabled via terraform I guess cc @uolter

yep @gunzip is right :+1:  it makes much more sense to disable the autoswap before ...

**Comment by balanza** [comment]
Draft until we apply https://github.com/pagopa/io-infrastructure-live-new/pull/292

**Comment by gunzip** [comment]
it looks like that one changes autoswap on io-functions-app not io-functions-assets

**Pull request merged**